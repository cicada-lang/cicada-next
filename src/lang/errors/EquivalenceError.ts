import { LangError } from "./LangError.js"

export class EquivalenceError extends LangError {
  trace: Array<string> = []

  constructor(public message: string) {
    super(message)
  }
}
