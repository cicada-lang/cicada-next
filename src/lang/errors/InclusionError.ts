import { LangError } from "./LangError.js"

export class InclusionError extends LangError {
  trace: Array<string> = []

  constructor(public message: string) {
    super(message)
  }
}
