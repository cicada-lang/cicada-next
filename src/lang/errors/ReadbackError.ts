import { LangError } from "./LangError.js"

export class ReadbackError extends LangError {
  constructor(public message: string) {
    super(message)
  }
}
