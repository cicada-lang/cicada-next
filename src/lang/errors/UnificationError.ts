import { LangError } from "./LangError.js"

export class UnificationError extends LangError {
  trace: Array<string> = []

  constructor(public message: string) {
    super(message)
  }
}
