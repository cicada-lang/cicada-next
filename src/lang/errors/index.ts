export { ParsingError } from "@cicada-lang/partech"
export * from "./AssertionError.js"
export * from "./ElaborationError.js"
export * from "./EquivalenceError.js"
export * from "./ErrorReport.js"
export * from "./EvaluationError.js"
export * from "./InclusionError.js"
export * from "./InternalError.js"
export * from "./LangError.js"
export * from "./ReadbackError.js"
export * from "./UnificationError.js"
