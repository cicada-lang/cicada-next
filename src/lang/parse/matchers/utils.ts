import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
