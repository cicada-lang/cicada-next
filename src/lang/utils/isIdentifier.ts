const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
