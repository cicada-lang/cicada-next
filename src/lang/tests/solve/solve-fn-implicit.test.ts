import { expect, test } from "bun:test"
import * as Errors from "../../errors"
import { expectCodeToFail, runCode } from "../utils"

test("solve FnImplicit", async () => {
  await expectCodeToFail(`

solve (x: String) {
  (implicit A: Type, a: A) => x =
  (implicit A: Type, a: A) => "abc"
}

`)
})

test("solve FnImplicit -- 2", async () => {
  await expectCodeToFail(`

solve (x: String) {
  (implicit A: Type, implicit B: Type, a: A, b: B) => x =
  (implicit A: Type, implicit B: Type, a: A, b: B) => "abc"
}

`)
})

test("solve FnImplicit -- alpha equivalence", async () => {
  const output = await runCode(`

solve () {
  (implicit A: Type, a: A) => a = (implicit B: Type, b: B) => b
}

`)

  expect(output).toMatchSnapshot()
})

test("solve FnImplicit -- deepWalk", async () => {
  const output = await runCode(`

solve (T: Type, f: (implicit _: Trivial) -> Type) {
  f = (implicit _) => T
  T = String
}

`)

  expect(output).toMatchSnapshot()
})

test("solve FnImplicit -- occur", async () => {
  await expectCodeToFail(
    `
  
solve (x: (implicit _: Type) -> Type) {
  x = (implicit y: Type) => x(implicit y)
}

`,
    Errors.ElaborationError,
  )
})
