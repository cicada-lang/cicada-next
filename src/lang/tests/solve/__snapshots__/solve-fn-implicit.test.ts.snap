// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`solve FnImplicit -- alpha equivalence 1`] = `"{}"`;

exports[`solve FnImplicit -- deepWalk 1`] = `
"{
  T: String,
  f: (implicit _) => String
}"
`;
